lib/realtime/vehicle.ex

```diff
@@ -64,20 +76,122 @@ defmodule Realtime.Vehicle do
+    timestamp = json["vehicle"]["timestamp"]
+    direction_id = json["vehicle"]["trip"]["direction_id"]
+    route_id = json["vehicle"]["trip"]["route_id"]
+    trip_id = json["vehicle"]["trip"]["trip_id"]
+    current_stop_status = decode_current_status(json["vehicle"]["current_status"])
+    stop_id = json["vehicle"]["stop_id"]
+
+    stop_times_on_trip =
+      try do
+        stop_times_on_trip_fn.(trip_id)
+      catch
+        # Handle Gtfs server timeouts gracefully
+        :exit, _ ->
+          []
+      end
+
+    {percent_of_the_way_to_next_timepoint, next_timepoint_stop_time} =
+      percent_of_the_way_to_next_timepoint(stop_times_on_trip, stop_id)
+
     %__MODULE__{
-      id: json["id"],
-      label: json["vehicle"]["vehicle"]["label"],
-      timestamp: json["vehicle"]["timestamp"],
-      direction_id: json["vehicle"]["trip"]["direction_id"],
-      route_id: json["vehicle"]["trip"]["route_id"],
-      trip_id: json["vehicle"]["trip"]["trip_id"],
-      current_status: decode_current_status(json["vehicle"]["current_status"]),
-      stop_id: json["vehicle"]["stop_id"]
+      id: id,
+      label: label,
+      timestamp: timestamp,
+      direction_id: direction_id,
+      route_id: route_id,
+      trip_id: trip_id,
+      current_stop_status: current_stop_status,
+      stop_id: stop_id,
+      current_timepoint_status:
+        timepoint_status(current_stop_status, next_timepoint_stop_time, stop_id),
+      timepoint_id: next_timepoint_id(next_timepoint_stop_time),
+      percent_of_the_way_to_timepoint: percent_of_the_way_to_next_timepoint
+    }
+  end
+
+  @spec percent_of_the_way_to_next_timepoint([StopTime.t()], Stop.id()) :: {
+          non_neg_integer(),
+          StopTime.t() | nil
+        }
+  def percent_of_the_way_to_next_timepoint([], _stop_id), do: {0, nil}
+
+  def percent_of_the_way_to_next_timepoint(stop_times, stop_id) do
+    {past_stop_times, current_stop_time, future_stop_times} =
+      split_stop_times(stop_times, stop_id)
+
+    next_timepoint_stop_time =
+      Enum.find([current_stop_time | future_stop_times], &is_a_timepoint?(&1))
+
+    {
+      percent_of_the_way(
+        count_to_timepoint(past_stop_times),
+        count_to_timepoint(future_stop_times)
+      ),
+      next_timepoint_stop_time
     }
   end
 
+  @spec split_stop_times([StopTime.t()], Stop.id()) ::
+          {[StopTime.t()], StopTime.t() | nil, [StopTime.t()]}
+  defp split_stop_times(stop_times, stop_id) do
+    stop_times
+    |> split_on_stop_id(stop_id)
+    |> past_current_future()
+  end
+
+  @spec split_on_stop_id([StopTime.t()], Stop.id()) :: {[StopTime.t()], [StopTime.t()]}
+  defp split_on_stop_id(stop_times, stop_id),
+    do: Enum.split_while(stop_times, &(&1.stop_id != stop_id))
+
+  @spec past_current_future({[StopTime.t()], [StopTime.t()]}) ::
+          {[StopTime.t()], StopTime.t() | nil, [StopTime.t()]}
+  defp past_current_future({past, [current | future]}), do: {Enum.reverse(past), current, future}
```

Comment:
I don't think `current` needs to be split out from `future`. If the vehicle is `:in_transit_to`, then the current stop is in the future.

Right now, the only place one is used without the other is in the counting, but current should be included in future there to so that a bus whose next stop is a timepoint counts to that timepoint instead of to the next.

Reply:
This percentage calculation doesn't worry about in transit to vs at. Including the current in next would make the result wrong when it is at a station (see the 40% test), and when it's in transit as well, because it will give the percentage value for the previous stop. Let's see how this level of granularity works in practice. If we want more position detail we can always adjust the calculation in the future to take in transit vs at into account.

Reply:
Discussed in person.

Reply:
👍 updated